File: backend/api/invites.py
from flask import Blueprint, request, jsonify
from adapters.firebase_adapter import verify_id_token, send_invite, respond_invite, invites, accept_invite
from adapters.myError import MyError

invites_blueprint = Blueprint('invites', __name__)

@invites_blueprint.route('/api/invites/<int:user_id>', methods=['POST'])
def send_invite(user_id):
    token = request.headers.get('authorization')
    sender_id = ""

    # Verify token and get user ID from it
    try:
        decoded_token = verify_id_token(token,"Unauthorized request")
        sender_id = decoded_token['uid']
    except MyError as e:
        return jsonify({"error": e.message}),e.number

    send_invite(sender_id,user_id)

    return jsonify({'message': 'Invitation sent successfully'}), 200

@invites_blueprint.route('/api/invites/<int:user_id>', methods=['GET'])
def get_invites(user_id):
    token = request.headers.get('authorization')

    # Verify token and get user ID from it
    try:
        decoded_token = verify_id_token(token,"Unauthorized request")
        incoming,outgoing = invites(user_id)
        return jsonify({"incoming": incoming, "outgoing": outgoing}), 200
    except MyError as e:
        return jsonify({"error": e.message}),e.number

@invites_blueprint.route('/api/invites/<int:user_id>/accept', methods=['POST'])
def accept_invite(user_id):
    token = request.headers.get('authorization')

    try:
        # Verify token and get user ID from it
        decoded_token = verify_id_token(token,"Unauthorized request")
        recipient_id = decoded_token['uid']
        #Accept invite logic
        respond_invite(recipient_id,user_id,'accepted')
        add_friend(user_id,recipient_id)
    except MyError as e:
        return jsonify({"error": e.message}),e.number

    return jsonify({'message': 'Invitation accepted successfully'}), 200

@invites_blueprint.route('/api/invites/<int:user_id>/decline', methods=['POST'])
def decline_invite(user_id):
    token = request.headers.get('authorization')

    try:
        # Verify token and get user ID from it
        decoded_token = verify_id_token(token,"Unauthorized request")
        recipient_id = decoded_token['uid']
        #Decline invite logic
        respond_invite(recipient_id,user_id,'declined')
    except MyError as e:
        return jsonify({"error": e.message}),e.number

    return jsonify({'message': 'Invitation declined successfully'}), 200

File: backend-dart/bin/routes/invites.dart
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';
import 'dart:convert';

import '../adapter/firebaseAdapter.dart';
import 'responses.dart';

class Invites {

  static Future<Response> sendInvite(Request request, String userId) async {
    final userData = await FirebaseAdapter.verifyHeader(request.headers['authorization']);

    if(userData.$2 != 200) {
      return error(userData.$2,userData.$1);
    }

    await FirebaseAdapter.sendInvite(userData.$1,userId);

    return Response.ok('Invitation sent successfully');
  }

  static Future<Response> getInvites(Request request,String userId) async {
    final userData = await FirebaseAdapter.verifyHeader(request.headers['authorization']);

    if(userData.$2 != 200) {
      return error(userData.$2,userData.$1);
    }

    (Map<String,dynamic>,Map<String,dynamic>) data = await FirebaseAdapter.getInvites(userData.$1);

    return Response.ok(jsonEncode({
      'incoming': data.$1,
      'outgoing': data.$2,
    }), headers: {'Content-Type': 'application/json'});

  }

  static Future<Response> declineInvite(Request request, String userId) async {
    final userData = await FirebaseAdapter.verifyHeader(request.headers['authorization']);

    if(userData.$2 != 200) {
      return error(userData.$2,userData.$1);
    }

    await FirebaseAdapter.declineInvite(userId,userData.$1);
    
    return Response.ok('Invitation declined successfully');
  }

  static Future<Response> acceptInvite(Request request, String userId) async {
    final userData = await FirebaseAdapter.verifyHeader(request.headers['authorization']);

    if(userData.$2 != 200) {
      return error(userData.$2,userData.$1);
    }

    await FirebaseAdapter.acceptInvite(userId,userData.$1);

    return Response.ok('Invitation accepted successfully');
  }

}

final invitesRouter = Router()
  ..post('/<user_id>', Invites.sendInvite)
  ..get('/<user_id>', Invites.getInvites)
  ..post('/<user_id>/decline', Invites.declineInvite)
  ..post('/<user_id>/accept', Invites.acceptInvite);